compass/landice/tests/mismipplus/setup_mesh.py

```diff
@@ -0,0 +1,301 @@
+from mpas_tools.planar_hex import make_planar_hex_mesh
+
+from compass.model import make_graph_file
+from compass.step import Step
+
+
+class SetupMesh(Step):
+    """
+    A step for creating a mesh and initial condition for MISMIP+ test cases
+
+    Attributes
+    ----------
+    resolution : int
+        The resolution of the test case, as defined configuration file at the
+        when `compass setup` is run.
+
+    nx : int
+        The number of cells in the x direction
+
+    ny : int
+        The number of cells in the y direction
+
+    dc : int
+        The distance in meters between adjacent cell centers.
+
+    """
+    def __init__(self, test_case, resolution):
+        """
+        Create the step
+
+        Parameters
+        ----------
+        testcase : compass.TestCase
+            The test case this step belongs to
+
+        resolution : int
+            The distance between horizontal grid points
+        """
+        # Defined resolutions, the value of `resolution` in the configuration
+        # file needs to a key in the dictionary, or added manually.
+        resolution_params = {'8km': {'nx': 73, 'ny': 14, 'dc': 9237.604307},
+                             '4km': {'nx': 144, 'ny': 24, 'dc': 4618.802154},
+                             '2km': {'nx': 284, 'ny': 44, 'dc': 2309.401077},
+                             '1km': {'nx': 566, 'ny': 84, 'dc': 1154.700538}}
+
+        # With the defined resolution, make the associated key for the param
+        # dictionary (neccessary b/c unstructured hex meshes)
+        key = f'{resolution}km'
+
+        # Ensure the resolution passed in the configuration file is defined
+        # within the parameter dictionary. If this is not the case, either:
+        #   1. choose an already defined resolution
+        #   2. add the appropriate parameter values to the dict.
+        assert key in resolution_params
+
+        super().__init__(test_case=test_case,
+                         name=f'{key}_mesh_gen',
+                         subdir=f"{key}/mesh_gen")
+
+        # Files to be created as part of the this step
+        for filename in ['mpas_grid.nc', 'graph.info', 'landice_grid.nc']:
+            self.add_output_file(filename=filename)
+
+        # unpack the mesh parameters for the given resolution as set as
+        # attributes
+        self.nx, self.ny, self.dc = resolution_params[key].values()
+
+        self.resolution = resolution
+
+    def run(self):
+        """
+        Run this step of the test case
+        """
+        logger = self.logger
+        config = self.config
+
+        section = config['mesh']
+
+        # read the resolution from the .cfg file at runtime
+        resolution = section.getint('resolution')
+
+        # check if the resolution has been changed since the `compass setup`
+        # command was run
+        if self.resolution != resolution:
+            warnings.warn(f'Resolution was set at {self.resolution:2d}km when'
```

Comment:
Your right, it probably makes more sense to raise an error. This version of the code raises a warning and then ignores the changes to resolution made after the time `compass setup` was run. But I could definitely see someone getting confused as to why the their resolution change didn't work. The warning will go to the log file, and it would frustrating to have to dig through that to figure out why the resolution change didn't work.